### ISSUE TITLE
Feat: Manage user membership for spaces of type OPEN

### ISSUE THREAD

**Issue by grignards** [tracim/tracim#3538]
## Feature description and goals

The `OPEN` space access type allows a user to discover and join a space of this type by himself.

### Create an API POST `/api/users/<user_id>/workspaces`

- [x] accepts a `{workspace_id: 42}` object
- [x] accepts only id of `OPEN` workspaces, returns 400 if not
- [x] adds the user as member with the defined default role (see #3601).
- [x] if already member, returns 400 with `RoleAlreadyExistError` exception

## Prerequisites

- #3536
- #3601 

### Which (default) role do the users get?

Possibilities:
  - a hard-coded value (`contributor`?)
  - a default value per instance: set in the backend config file
  - a default value per space: set during the space creation

**Decision**: we will allow a default role per space as use-cases for this will arrive soon, see #3601.

### Are the users automatically added when creating an `OPEN` space?

Or is this a separate user action (which would allow to create a space before "publishing" it)?

**Decision**:  by default Tracim won't do anything when a space is created. Automated membership of all users will be implemented separately, see #3604

### Can the OPEN access type be disabled (by configuration)?

If yes should we handle the `OPEN` access type as a plugin?
If we go this way (customizable spaces types) we should provide an API access to the possible space access types (probably through the system config API).

**Decision**: yes it will be possible to disable it (and `ON_REQUEST`). See #3602 for details and implementation.

**Comment by gbonaspetti** [comment]
> Which (default) role do the users get?
Possibilities:
a hard-coded value (contributor?)
a default value per instance: set in the backend config file
a default value per space: set during the space creation

In a first moment I would choose the first or the second option (maybe with a slight preference for the second, because I can see cases where the client may want "Reader" as the default value). The third one seems to me an improvement for the future.

> Are the users automatically added when creating an OPEN space?
Or is this a separate user action (which would allow to create a space before "publishing" it)?

Same thing, I tend to say: first option, considering the second as an improvement for next versions.

>Can the OPEN access type be disabled (by configuration)?
If yes should we handle the OPEN access type as a plugin?
If we go this way (customizable spaces types) we should provide an API access to the possible space access types (probably through the system config API).

I don't have much of an opinion on this question, as I don't know the difference in development time. If it's not too much, maybe it's a good idea to be able to deactivate.

**Comment by grignards** [comment]
> In a first moment I would choose the first or the second option (maybe with a slight preference for the second, because I can see cases where the client may want "Reader" as the default value). The third one seems to me an improvement for the future.

I agree, my preference would be for the second option has it is almost the same development time and ensures a bit of flexibility to Tracim administrators. The third one is probably the best "user-wise" but also the more time-consuming implementation.

**Comment by grignards** [comment]
> Same thing, I tend to say: first option, considering the second as an improvement for next versions.

I also agree as having a real "not-published/draft" space is more complex than that: normal users will be able to see (and join) an OPEN space as soon as it is created even with this feature.

**Comment by inkhey** [comment]
> ### Which (default) role do the users get?
> 
> Possibilities:
> 
>     * a hard-coded value (`contributor`?)
> 
>     * a default value per instance: set in the backend config file
> 
>     * a default value per space: set during the space creation
> 

I do suggest a backend config value like for `user.default_profile`

> ### Are the users automatically added when creating an `OPEN` space?
> 
> Or is this a separate user action (which would allow to create a space before "publishing" it)?

I'm not sure to understand this one, but i don't see the point of allowing rejoin/leave workspace by api and also automatically add them to open space. 
I do prefer we do not automatically add user to workspace but directly allow user to see these space where user can ask for join (which doesn't does same for open and request space).


> ### Can the OPEN access type be disabled (by configuration)?
> 
> If yes should we handle the `OPEN` access type as a plugin?
> If we go this way (customizable spaces types) we should provide an API access to the possible space access types (probably through the system config API).

It depends how we expect things to work in both public instance like MonTracim but also in organization instance. Do we have case we want user to both:
- create their own workspace
- do not create open/request workspaces

I do think the problematic case is just with trusted-users (admin are admin, so disabling feature doesn't make much sense).
A parameter limiting the type of workspace a trusted user can create seem to me to most logical feature here.

**Comment by grignards** [comment]
> Are the users automatically added when creating an OPEN space?

There has been the idea of separating the creation of an `OPEN` space and the moment the users are added to it so that space managers can choose a different time to add them.

> I do prefer we do not automatically add user to workspace but directly allow user to see these space where user can ask for join (which doesn't does same for open and request space).

The goal for our client is to make some information directly available to all users as soon as they register. Once they have discovered what they are interested in they can leave some spaces which have no interest for them. Allowing them to re-join a space is to allow them to change their minds.

**Comment by grignards** [comment]
>I do think the problematic case is just with trusted-users (admin are admin, so disabling feature doesn't make much sense).

Even for admins it can make sense as admins in Tracim are not necessarily the same people as "system admins" who have access to the configuration files. 

> A parameter limiting the type of workspace a trusted user can create seem to me to most logical feature here.

Yes this could be an idea, a bit more complex than only enabling/disabling support entirely. I'm also unsure it is necessary for "private" instances as trusted users can already invite anybody to their spaces, so I think the main goal is to make the differences between OPEN/ON_REQUEST/CONFIDENTIAL spaces clear to trusted users/administrators.

**Comment by inkhey** [comment]
> > 
> 
> There has been the idea of separating the creation of an `OPEN` space and the moment the users are added to it so that space managers can choose a different time to add them.

it's unclear to me what that's mean. Currently only the creator is added to the space newly created, you have to manually add other user. I don't see any advantage to not include the creator into his own space at start, and if we don't add a feature that automatically add multiple user into a space, there is no need to change this.
@grignards can you clarify this point.

**Comment by inkhey** [comment]
> The goal for our client is to make some information directly available to all users as soon as they register. Once they have discovered what they are interested in they can leave some spaces which have no interest for them. Allowing them to re-join a space is to allow them to change their minds.

If this is the idea and this is a specific behavior for some client (some other should want something different), it may be more easier to do this in frontend, with or without a pop up allowing them or not to "discover all" ?

**Comment by lebouquetin** [comment]
I think we need to have a discussion about the behavior on a user fonctionality point of view.

**Comment by grignards** [comment]
> > There has been the idea of separating the creation of an `OPEN` space and the moment the users are added to it so that space managers can choose a different time to add them.
> 
> it's unclear to me what that's mean. Currently only the creator is added to the space newly created, you have to manually add other user. I don't see any advantage to not include the creator into his own space at start, and if we don't add a feature that automatically add multiple user into a space, there is no need to change this.
> @grignards can you clarify this point.

One of the characteristics of `OPEN` spaces is that all users are member of them initially. We've had two ideas for implementing this:

- either all users directly become a member when a space of this type is created
- OR an easy way to add them is provided to trusted users/admins after creation

Those 2 implementations do not change the fact that the owner of a space is automatically member of it.

**Comment by inkhey** [comment]
> > > There has been the idea of separating the creation of an `OPEN` space and the moment the users are added to it so that space managers can choose a different time to add them.
> > 
> > 
> > it's unclear to me what that's mean. Currently only the creator is added to the space newly created, you have to manually add other user. I don't see any advantage to not include the creator into his own space at start, and if we don't add a feature that automatically add multiple user into a space, there is no need to change this.
> > @grignards can you clarify this point.
> 
> One of the characteristics of `OPEN` spaces is that all users are member of them initially. We've had two ideas for implementing this:
> 
>     * either all users directly become a member when a space of this type is created
> 
>     * OR an easy way to add them is provided to trusted users/admins after creation
> 
> 
> Those 2 implementations do not change the fact that the owner of a space is automatically member of it.

After some check:
- admin know the list of all members.
- trusted_users as way to know it but this is less easy, allowing such feature will be less "secure" in a way.

Also in a public instance context like mon tracim, i do see 2 different case for this:
- OPEN space "default" space with membership for all added automatically at user creation/workspace creation.
- OPEN space "available" which we can request to get into but don't need acceptation (auto-accept)

**Comment by grignards** [comment]
> Also in a public instance context like mon tracim, i do see 2 different case for this:
> 
>     * OPEN space "default" space with membership for all added automatically at user creation/workspace creation.
> 
>     * OPEN space "available" which we can request to get into but don't need acceptation (auto-accept)

I have mixed feelings whether OPEN/ON_REQUEST spaces are useful for mon.tracim.fr or other "public" instances.

**Comment by inkhey** [comment]
Just created a table to clarify all the possibility and which one and which one doesn't make sense:

Currently, in specification, we do have something like: 

|              | auto-subscribe | visible | auto-accept | description                                 |
|--------------|------------------|------------|------------------|---------------------------------------------|
| OPEN(current)         | ✅                | ✅          | ✅                |                                             |
| ON_REQUEST   | ❌                | ✅          | ❌                |                                             |
| CONFIDENTIAL | ❌                | ❌          | ❌                |                                             |

And these are the case we do not handle:

|              | auto-subscribe | visible | auto-accept | description                                 |
|--------------|------------------|------------|------------------|---------------------------------------------|
| OTHER1       | ❌                | ✅          | ✅                | this one make sense to me   |          
| OTHER2       | ✅                | ❌          | ❌                | not really useful                   |
| OTHER3       | ✅                | ✅          | ❌                | not really uselful                   |
| OTHER4       | ✅                | ❌          | ✅                |  impossible to accept if not visible |
| OTHER5       | ❌                | ❌          | ✅                |  impossible to accept if not visible |

The case OTHER1 make sense to me, allowing adding all user by ui is not same as autosubscribe new user to workspace, so should we:
- choose only one implementation between OPEN(current) and OTHER1  for "OPEN" ? We can add a "add all button" if we choose OTHER1 implementation but this will not be the same thing.
- find a way to handle both case, it can be by add new type or add other configuration parameter for open or on_request ?

**Comment by lebouquetin** [comment]
- Default role : should be parameterized by space. Example: I have a forum section (where default role should be contributor) and a documentation section (where default role should be reader). By the way, the default role will allow to prefill invite screens (for admin + dashboard)
- auto-subscription:
  - on space creation: there is no need for implementing something except a "user experience that allow to directly choose members / all members"
  - on user joining tracim: this may/should be done through a backend plugin based on events. This will allow any type of auto-subscription)
- disabling an OPEN shared space may be a nice feature; I this this is not related to OPEN type but on an action decided to temporary make a space disabled/readonly
- visibility seems related to type of space: if not confidential, then visibility is the standard.

Consider that joining/leaving a space is not related to its type (PUBLIC spaces may be leaved if they are of no interest for you, even if they are public. Same thing for OPEN / ON DEMAND - except in last case that you can't join as you want)

**Comment by lebouquetin** [comment]
- Creating an OPEN/ON_REQUEST space make it visible for all users. It is acceptable at the moment
- Creating an OPEN/ does not invite automatically users (this may/will be done through a backend plugin)
- Creating a user account does not invite automatically the user in OPEN spaces (this may/will be done through a backend plugin)

**Comment by inkhey** [comment]
should be #3601 a requirement ?

**Comment by inkhey** [comment]
-> if already member, returns 400

Should be RoleAlreadyExistError error, error code `3008`

**Comment by grignards** [comment]
> should be #3601 a requirement ?

#3604 would be a better candidate for this requirement, I'll add it.

**Comment by inkhey** [comment]
accepts only id of OPEN workspaces, returns 400 if not -> we need clarifiyng the error code here too.

**Comment by lebouquetin** [comment]
My feeling is that joining an open space should be like making a request which auto validates. Otheewise processes are differents for each type of space, which is more optimized but more difgivult to understand the global behavior of the software

**Comment by lebouquetin** [comment]
By the way it makes endpoints more difficult to understand how/when to use them

**Comment by raphj** [comment]
How about using this address for this endpoint: `/api/users/<user_id>/workspaces/join`?

**Comment by grignards** [comment]
> By the way it makes endpoints more difficult to understand how/when to use them

I did consider it, but the actions are inherently different (for `ON_REQUEST` the call creates a resource for the subscription, but does not make the space available in `workspaces`, for `OPEN` it would not create a resource for the subscription but make the space available…). I agree it is complex but am not really convinced putting both in the same API will really simplify its usage, for me it would only help initially (one endpoint), but not after (this endpoint's behavior is complex).

**Comment by lebouquetin** [comment]
> it would only help initially (one endpoint), but not after (this endpoint's behavior is complex).

- it *would* help using the API
- it *would* complexify the endpoint behavior 
- it *would* unify workflow (so simplify global comprehension of "how tracim works")

If you consider 3 routes is the best solution, the route naming should be more explicit because each route is for a specific case eg ` POST /api/users/<user_id>/workspaces` should be something like  `POST /api/users/<user_id>/open_workspaces`, no?

**Comment by grignards** [comment]
> If you consider 3 routes is the best solution, the route naming should be more explicit because each route is for a specific case eg ` POST /api/users/<user_id>/workspaces` should be something like `POST /api/users/<user_id>/open_workspaces`, no?

`open_workspaces` route does not exist, I chose the "classic" REST way for creating a resource: POST-ing in the collection (since it adds a workspace in `GET /api/users/<user_id>/workspaces`).

BTW having one endpoint for both (or even one for all `/api/workspaces/{workspace_id}/members/{user_id}` could work) would better be suited with a `PENDING` member role that I did not choose as I think it would be error prone security-wise (it is more intuitive that every role has access, rather than having a role which does not give any access).

**Comment by inkhey** [comment]
What about conserving:
```
API GET /api/users/<user_id>/workspace_subscriptions
API GET /api/workspaces/<workspace_id>/subscriptions
API PUT /api/workspaces/<workspace_id>/subscriptions/<user_id>/accept
API PUT /api/workspaces/<workspace_id>/subscriptions/<user_id>/reject
```
but merging both : 
```
API POST /api/users/<user_id>/workspaces
API PUT /api/users/<user_id>/workspace_subscriptions
```
into a
`API POST /api/users/<user_id>/workspaces`
that return either 400 or 204 (if open workspace) or 202 (if request workspace) ?

**Comment by grignards** [comment]
> into a
> `API POST /api/users/<user_id>/workspaces`
> that return either 400 or 204 (if open workspace) or 202 (if request workspace) ?

This has been discussed just before: one api is better for quick understanding but the api's behavior is more complex than having two (since you merge the two apis behaviors). It also make the subscription's creation implicit.

**Comment by grignards** [comment]
> @grignards I not understand why no test done here. This is visible in MR `Manual, quality tests have not been done: As the API is new testing this PR should be done with #3587` and issue related is for Tracim 3.2.1.
> 
> Please add a comment here to explain what is necessary to test in 3.2.0 because if test can just be done when `#3587` is fix, we can't do anything here.
> Thanks in advance.

The front-end currently doesn't use the APIs which have been developed in this issue (they will use them when #3587 is implemented and it was initially in 3.2.0 which is why I wrote to wait for it). So testing here could only be done through curl/manually created HTTP requests which has little value as they will provide the same tests as the automated back-end tests written during the development.

A possible test would be:

With Firefox:

1. create an OPEN space with the front-end and take note of its id.
2. create a user with the front-end
3. log as the created user
4. open the developer toolbox and reload the page
5. go to the network tab and modify an API request (eg. `http://localhost:7999/api/users/2/config`):
  - change the request method to `POST`
  - change the request path to `http://localhost:7999/api/users/2/workspaces`
  - set the request body to be

```json
{
  "workspace_id": id
}
```
  - send the request
6. the workspace should be added in the sidebar.

**Comment by PhilippeAccorsi** [comment]
Thanks. Thats working good.
